### PR TITLE
[SPARK] Migrated 'shared' module to produce Scala 2.12 and Scala 2.13 variants

### PR DESCRIPTION
**Pull request by d-m-h**
### Problem

Spark 3.2.x, 3.3.x, 3.4.x, and 3.5.x are compiled using Scala 2.12 and Scala 2.13. Due to a change in the Scala Collections API in Scala 2.13, NoSuchMethodErrors are thrown when running the openlineage-spack connector in an Apache Spark runtime when the runtime was compiled using Scala 2.13.

Relates to: #2303 

### Solution

This PR is the **fourth** of several PRs to support producing Scala 2.12 and Scala 2.13 variants of the OpenLineage Spark integration.

In this PR, we migrate the 'shared' module to use the refactored Gradle plugins. We also split some tests into Scala 2.12 and Scala 2.13 specific versions.

The module now produces 3 JARS; one default one, one compiled using the Scala 2.12 variant of Apache Spark, and one compiled using the Scala 2.13 variant of Apache Spark.

### Checklist

- [x] You've [signed-off](https://github.com/OpenLineage/OpenLineage/blob/main/why-the-dco.md) your work
- [x] Your pull request title follows our [guidelines](https://github.com/OpenLineage/OpenLineage/blob/main/CONTRIBUTING.md#creating-pull-requests)
- [x] Your change contains a [small diff](https://kurtisnusbaum.medium.com/stacked-diffs-keeping-phabricator-diffs-small-d9964f4dcfa6) and is self-contained
- [x] Your comment includes a one-liner for the changelog about the specific purpose of the change (_if necessary_)
- [x] You've added a [header](https://github.com/OpenLineage/OpenLineage/tree/main/.github/header_templates.md) to source files (_if relevant_)

----
SPDX-License-Identifier: Apache-2.0\
Copyright 2018-2023 contributors to the OpenLineage project

### CODE REVIEW

File: integration/spark/shared/src/test/java/io/openlineage/spark/agent/lifecycle/LogicalPlanSerializerTest.java [hunk 1]
```diff
@@ -269,6 +206,7 @@ void testSerializeInsertIntoHadoopPlan()
   }
 
   @Test
+  @Disabled("This one is throwing a ClassCastException and needs to be fixed")
```
[the remark this answers is not in the document]
Comment (d-m-h) [reply integration/spark/shared/src/test/java/io/openlineage/spark/agent/lifecycle/LogicalPlanSerializerTest.java:209]: If I recall correctly, I was experiencing the failure with Scala 2.12 as well. I think it has to do with the version of Spark x BigQuery used - but I may be confusing it with something else.

File: integration/spark/shared/src/test/java/io/openlineage/spark/agent/OpenLineageRunEventTest.java [hunk 2]
```diff
@@ -36,8 +36,7 @@ public void testSerializeRunEvent() throws IOException, URISyntaxException {
     ZonedDateTime dateTime = ZonedDateTime.parse("2021-01-01T00:00:01.000000000+00:00[UTC]");
     OpenLineage ol =
         new OpenLineage(
-            new URI(
-                "https://github.com/OpenLineage/OpenLineage/tree/0.2.3-SNAPSHOT/integration/spark"));
+            new URI("https://github.com/OpenLineage/OpenLineage/tree/0.2.3-SNAPSHOT//spark"));
```
[the remark this answers is not in the document]
Comment (d-m-h) [reply integration/spark/shared/src/test/java/io/openlineage/spark/agent/OpenLineageRunEventTest.java:39]: Perhaps a find-and-replace error. I'll revert this.

File: integration/spark/shared/build.gradle [hunk 3]
```diff
@@ -1,96 +1,201 @@
...
+    testScala212Implementation("org.apache.kafka:kafka-clients:${kafkaClientsVersion}")
+    testScala212Implementation("org.assertj:assertj-core:${assertjVersion}")
+    testScala212Implementation("org.awaitility:awaitility:${awaitilityVersion}")
+    testScala212Implementation(platform("org.junit:junit-bom:${junit5Version}"))
+    testScala212Implementation("org.junit.jupiter:junit-jupiter")
+    testScala212Implementation("org.junit.jupiter:junit-jupiter-params")
+    testScala212Implementation("org.mockito:mockito-core:${mockitoVersion}")
+    testScala212Implementation("org.mockito:mockito-inline:${mockitoVersion}")
+    testScala212Implementation("org.mockito:mockito-junit-jupiter:${mockitoVersion}")
+    testScala212Implementation("org.mock-server:mockserver-client-java:${mockserverVersion}") {
```
[the remark this answers is not in the document]
Comment (d-m-h) [reply integration/spark/shared/build.gradle:139]: I honestly contemplated doing this myself, but I didn't want to introduce too many changes. (It's why I kept the test fixtures stuff in the modules)

File: integration/spark/shared/src/test/java/io/openlineage/spark/agent/lifecycle/LogicalPlanSerializerTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (pawel-big-lebowski) [reply integration/spark/shared/src/test/java/io/openlineage/spark/agent/lifecycle/LogicalPlanSerializerTest.java:209]: Yeah, big query connector for shared was compiled with Scala `2.11` and it seems as if the test was only passing under this assumption. Would be great if you were able to fix this or give a reason why it is difficult to solve.

[the remark this answers is not in the document]
Comment (d-m-h) [reply integration/spark/shared/src/test/java/io/openlineage/spark/agent/lifecycle/LogicalPlanSerializerTest.java]: As discussed in Slack, I removed this test.

**Review by pawel-big-lebowski — APPROVED** [review]

**Pull request merged**